Rewrite AsyncLoad with React.lazy (and simplify wpcalypso-async)

jsnajdr opened the pull request:
This PR is a rewrite of the `AsyncLoad` component that makes it use `React.lazy`. The old version was a class component with local state and effects that triggered the loading. That is now gone and the implementation is much simpler. And it's natively TypeScript.

I had to change the `wpcalypso-async` transform so that the argument passed to the `require` callback is not the `default` export of the loaded module, but the entire module object itself. You now need to dereference `.default` explicitly. This is to make it compatible with `React.lazy`, which expects an object with shape `{ default: ComponentType }`.

As a related change, I modified all usages of `asyncRequire` to use the version that returns a promise. Makes them look more like dynamic `import` calls.

Lastly, I removed the `async` option from the `wpcalypso-async` Babel transform -- everything is `async` now by default. We were not using the `async: false` version anywhere anyway. There's only the `ignore` option that makes the async loads into noops -- useful on server.

This is the first step in my "simplify and eliminate" strategy to remove the `wpcalypso-async` transform completely.

**How to test:**
Test that all async loaded components still work. And that the `asyncRequire` calls work, too. They are used for dev helpers in dev mode, and also for loading emoji libraries in Reader, on the `/tags/:tag` page.

jsnajdr committed:
wpcalypso-async Babel transform: remove the async option, be async by default

jsnajdr committed:
wpcalypso-async Babel transform: call requie callback with a module object, not the .default field

jsnajdr committed:
Rewrite asyncRequire usages to promise form

jsnajdr committed:
Rewrite AsyncLoad with React.lazy, and in TypeScript

jsnajdr commented on client/components/async-load/index.tsx:
```diff
@@ -0,0 +1,39 @@
+type AsyncLoadProps = {
+	placeholder?: ReactNode;
+	require: string;
+	[ key: string ]: unknown;
```
The type of `AsyncLoad` should be eventually improved, by specifying the component type that is supposed to be loaded from the async module, and to check the props agains the type.

In the current form, you can pass arbitrary component props to `AsyncLoad` and they won't be checked. That's the type behavior that was already there before this PR.

tyxla replied:
I think that's perfectly fine for the first pass 👍

tyxla commented on client/boot/common.js:
```diff
@@ -224,8 +224,8 @@ const configureReduxStore = ( currentUser, reduxStore ) => {
-		asyncRequire( 'calypso/lib/network-connection', ( networkConnection ) =>
-			networkConnection.init( reduxStore )
+		asyncRequire( 'calypso/lib/network-connection' ).then( ( networkConnection ) =>
+			networkConnection.default.init( reduxStore )
```
It's a bit of an inconvenience that we have to use `.default`. I wonder if this is something we can improve in the `asyncRequire()` function in the babel transform itself. Which won't be an issue if we're trying to get rid of the entire transform in the long term of course.

jsnajdr replied:
It's inconvenient, yes, but on the other hand, that's what native `import()` does. It also returns a module object where you need to access `.default` explicitly.

So I'm going in that direction: the `asyncRequire` call with eventuall become `import` without having change to anything else.

tyxla replied:
Sounds good 👍

tyxla commented on packages/babel-plugin-transform-wpcalypso-async/CHANGELOG.md:
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
+- Removed the `async` option and started transforming all usages to async loads by default
+- The `require` callback now gets the module object as a parameter, not just its default field (for compatibility with `React.lazy`)
```
Should these be marked as breaking changes? Even if it's just prefixing the log entries with "Breaking: " or "Breaking change:"

jsnajdr replied:
Good point, I added `Added` and `Removed` sections, according to the keepachangelog.com convention declared in the Changelog header.

tyxla commented on packages/babel-plugin-transform-wpcalypso-async/CHANGELOG.md:
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
```
If we're bumping to 2.0 in this PR, should we release those changes and add this under the new version? In other words, should we bump the version in the same PR where we update the CHANGELOG to be for 2.0?

jsnajdr replied:
There is one practical reason to bump the version now: it will clear the Babel cache. The version field is used to invalidate the cache key:

https://github.com/Automattic/wp-calypso/blob/trunk/build-tools/babel/babel-loader-cache-identifier/index.js#L23-L25

I ran into this when working on the PR: I modified the transform but the dev server was still using the old one, even after restarting, because it was using cached transpiled files.

tyxla replied:
Makes sense, thanks for elaborating!

tyxla reviewed (approved):
Both `asyncRequire` and `AsyncLoad` appear to work great 👍 

Code also looks good.

Great work 🚀

jsnajdr committed:
Edited the changelog to adhere to the keepachangelog convention

jsnajdr commented:
I'll do some more testing of the server bundle and SSR before deploying this. I'm not sure everything is 100% there.

tyxla commented:
> I'll do some more testing of the server bundle and SSR before deploying this. I'm not sure everything is 100% there.

I did some smoke testing on the server side FWIW (themes and plugins), but yeah, some additional testing can never be unnecessary.

jsnajdr committed:
wpcalypso-async Babel transform: promisify callback, throw errors in ignore mode

jsnajdr committed:
Convert dynamic imports into asyncRequire so that they can get ignored

noahtallen commented:
[The logs](https://teamcity.a8c.com/buildConfiguration/calypso_WpDesktop_DesktopE2ETests/11240771?buildTab=artifacts#%2Fresults) are noting an issue in the linux desktop app:

```sh
Fri, 17 Nov 2023 11:26:00 GMT [error] Failed to load resource: net::ERR_INSUFFICIENT_RESOURCES
Fri, 17 Nov 2023 11:26:00 GMT [error] Error: Loading CSS chunk 39675 failed.
(https://container-upbeat-mcnulty.calypso.live/calypso/evergreen/async-load-calypso-components-global-notices.53081c4bf3828ec7b218.min.css)
    at b.onerror.b.onload (https://container-upbeat-mcnulty.calypso.live/log-in:41:61826)
```

The error comes from `async-load-calypso-components-global-notices.53081c4bf3828ec7b218.min.css`, which could be related! Another thought is that electron is likely out of date.

jsnajdr commented:
> `net::ERR_INSUFFICIENT_RESOURCES`

This looks more like the client (Electron) is running out of resources -- too many opened files/sockets etc. Doesn't look like an error in Calypso, client or server. I restarted the tests.

jsnajdr commented:
> It's like the `React.lazy` callback was supposed to throw an error in SSR mode.

I fixed this by making the `React.lazy` callback throw an error right away, synchronously, in server mode (`ignore: true`). That's what the `ReactDomServer.renderToString` renderer expects: it runs synchronously, and it never waits for promises to resolve. It only checks which Suspense tasks are "pending" after the sync render finishes, and reports that they shouldn't be there in the first place.

To make this work, I had to further simplify the `asyncRequire`/`AsyncLoad` API to support only returning promises, and removed the callback version for good.
